spring-xd-messagebus-spi/src/main/java/org/springframework/xd/dirt/integration/bus/MessageBusSupport.java

```diff
@@ -898,7 +898,11 @@ private void revertDirectBindingIfNecessary(Binding binding) {
 
 		@Override
 		public int selectPartition(Object key, int partitionCount) {
-			return Math.abs(key.hashCode()) % partitionCount;
+			int hashCode = key.hashCode();
+			if (hashCode == Integer.MIN_VALUE) {
+				hashCode = 0;
+			}
+			return Math.abs(hashCode) % partitionCount;
```

Comment:
To clarify, my snippet above is what is currently in `MessageBusSupport.determinePartition()`.

And what I'm saying is that
* Modulo should be applied only once ideally
* and it's our responsibility to make sure it is applied (and correctly, see my point about being negative and less than partitionCount in absolute value)

==> this logic should go in MessageBusSupport (not be left to the selector strategy, where users could screw up)